native/dgebal.go

```diff
@@ -0,0 +1,227 @@
+	switch job {
+	default:
+		panic(badJob)
+	case lapack.None, lapack.Permute, lapack.Scale, lapack.PermuteScale:
+	}
+	checkMatrix(n, n, a, lda)
+	if len(scale) != n {
+		panic("lapack: bad length of scale")
+	}
+
+	ilo = 0
+	ihi = n - 1
+
+	if n == 0 || job == lapack.None {
+		for i := range scale {
+			scale[i] = 1
+		}
+		return ilo, ihi
+	}
+
+	bi := blas64.Implementation()
+	swapped := true
+
+	if job == lapack.Scale {
+		goto Scaling
+	}
+
+	// Permutation to isolate eigenvalues if possible.
+	//
+	// Search for rows isolating an eigenvalue and push them down.
+	for swapped {
+		swapped = false
+	RowLoop:
+		for i := ihi; i >= 0; i-- {
+			for j := 0; j <= ihi; j++ {
+				if i == j {
+					continue
+				}
+				if a[i*lda+j] != 0 {
+					continue RowLoop
+				}
+			}
+			// Row i has only zero off-diagonal elements in the
+			// block A[ilo:ihi+1,ilo:ihi+1].
+			scale[ihi] = float64(i)
+			if i != ihi {
+				bi.Dswap(ihi+1, a[i:], lda, a[ihi:], lda)
+				bi.Dswap(n, a[i*lda:], 1, a[ihi*lda:], 1)
+			}
+			ihi--
+			if ihi == 0 {
+				scale[0] = 1
+				return ilo, ihi
+			}
+			swapped = true
+			break
+		}
+	}
+	// Search for columns isolating an eigenvalue and push them left.
+	swapped = true
+	for swapped {
+		swapped = false
+	ColLoop:
+		for j := ilo; j <= ihi; j++ {
+			for i := ilo; i <= ihi; i++ {
+				if i == j {
+					continue
+				}
+				if a[i*lda+j] != 0 {
+					continue ColLoop
+				}
+			}
+			// Column j has only zero off-diagonal elements in the
+			// block A[ilo:ihi+1,ilo:ihi+1].
+			scale[ilo] = float64(j)
+			if j != ilo {
+				bi.Dswap(ihi+1, a[j:], lda, a[ilo:], lda)
+				bi.Dswap(n-ilo, a[j*lda+ilo:], 1, a[ilo*lda+ilo:], 1)
+			}
+			swapped = true
+			ilo++
+			break
+		}
+	}
+
+Scaling:
+	for i := ilo; i <= ihi; i++ {
+		scale[i] = 1
+	}
+
+	if job == lapack.Permute {
+		return ilo, ihi
+	}
+
+	// Balance the submatrix in rows ilo to ihi.
+
+	const (
+		// sclfac should be a power of 2 to avoid roundoff errors.
+		// Elements of scale are restricted to powers of sclfac,
+		// therefore the matrix will be only nearly balanced.
+		sclfac = 2
+		// factor determines the minimum reduction of the row and column
+		// norms that is considered non-negligible. It must be less than 1.
+		factor = 0.95
+	)
+	sfmin1 := dlamchS / dlamchP
+	sfmax1 := 1 / sfmin1
+	sfmin2 := sfmin1 * sclfac
+	sfmax2 := 1 / sfmin2
+
+	// Iterative loop for norm reduction.
+	conv := false
+	for !conv {
+		conv = true
+		for i := ilo; i <= ihi; i++ {
+			c := bi.Dnrm2(ihi-ilo+1, a[ilo*lda+i:], lda)
+			r := bi.Dnrm2(ihi-ilo+1, a[i*lda+ilo:], 1)
+			ica := bi.Idamax(ihi+1, a[i:], lda)
+			ca := math.Abs(a[ica*lda+i])
+			ira := bi.Idamax(n-ilo, a[i*lda+ilo:], 1)
+			ra := math.Abs(a[i*lda+ilo+ira])
+
+			// Guard against zero c or r due to underflow.
+			if c == 0 || r == 0 {
+				continue
+			}
+			g := r / sclfac
+			f := 1.0
+			s := c + r
+			for c < g && math.Max(f, math.Max(c, ca)) < sfmax2 && math.Min(r, math.Min(g, ra)) > sfmin2 {
```

Comment:
No no, definitely worse :-) The `GO TO 170` is attached to the condition like in the previous case _and_ the alignment of `IF` on line 336 is as if the body belonged to `IF` on line 334.